File: src/js/core/tests/Parity.test.js
// Cross-language parity tests. Reads the shared fixtures at
// tests/fixtures/parity/ and asserts that the JS implementation produces
// exactly the values committed there. The PHP test suite at
// src/php/core/tests/ParityTest.php reads the same fixtures and runs the
// equivalent assertions, so any divergence between the two implementations
// will surface as a failure on whichever side drifts.
//
// To regenerate the fixtures: `npm run fixtures:parity` (from repo root).

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';

import { Avatar, OptionsDescriptor, Style } from '../lib/index.js';
import { Prng } from '../lib/Prng.js';
import { Fnv1a } from '../lib/Prng/Fnv1a.js';
import { Mulberry32 } from '../lib/Prng/Mulberry32.js';
import { Number } from '../lib/Utils/Number.js';
import { Initials } from '../lib/Utils/Initials.js';
import { Color } from '../lib/Utils/Color.js';
import { ValidationError } from '../lib/Error/ValidationError.js';
import { CircularColorReferenceError } from '../lib/Error/CircularColorReferenceError.js';

const fixturesDir = join(import.meta.dirname, '..', '..', '..', '..', 'tests', 'fixtures', 'parity');

function loadFixture(relPath) {
  return JSON.parse(readFileSync(join(fixturesDir, relPath), 'utf8'));
}

describe('Parity / Fnv1a', () => {
  for (const entry of loadFixture('fnv1a.json')) {
    it(`hashes ${JSON.stringify(entry.input)}`, () => {
      assert.equal(Fnv1a.hash(entry.input), entry.hash);
      assert.equal(Fnv1a.hex(entry.input), entry.hex);
    });
  }
});

describe('Parity / Mulberry32', () => {
  for (const entry of loadFixture('mulberry32.json')) {
    it(`produces sequence for seed ${entry.seed}`, () => {
      const m = new Mulberry32(entry.seed);
      for (const expected of entry.sequence) {
        const float = m.nextFloat();
        const state = m.state();
        assert.equal(float, expected.float);
        assert.equal(state, expected.state);
      }
    });
  }
});

describe('Parity / Prng', () => {
  const fixture = loadFixture('prng.json');

  describe('getValue', () => {
    for (const c of fixture.getValue) {
      it(`seed=${JSON.stringify(c.seed)} key=${JSON.stringify(c.key)}`, () => {
        assert.equal(new Prng(c.seed).getValue(c.key), c.result);
      });
    }
  });

  describe('pick', () => {
    for (const [i, c] of fixture.pick.entries()) {
      it(`#${i} seed=${JSON.stringify(c.seed)} key=${JSON.stringify(c.key)}`, () => {
        assert.equal(new Prng(c.seed).pick(c.key, c.items), c.result);
      });
    }
  });

  describe('weightedPick', () => {
    for (const [i, c] of fixture.weightedPick.entries()) {
      it(`#${i} seed=${JSON.stringify(c.seed)} key=${JSON.stringify(c.key)}`, () => {
        assert.equal(new Prng(c.seed).weightedPick(c.key, c.weights), c.result);
      });
    }
  });

  describe('bool', () => {
    for (const [i, c] of fixture.bool.entries()) {
      it(`#${i} seed=${JSON.stringify(c.seed)} likelihood=${c.likelihood}`, () => {
        assert.equal(new Prng(c.seed).bool(c.key, c.likelihood), c.result);
      });
    }
  });

  describe('float', () => {
    for (const [i, c] of fixture.float.entries()) {
      it(`#${i} range=${JSON.stringify(c.range)}`, () => {
        assert.equal(new Prng(c.seed).float(c.key, c.range), c.result);
      });
    }
  });

  describe('integer', () => {
    for (const [i, c] of fixture.integer.entries()) {
      it(`#${i} range=${JSON.stringify(c.range)}`, () => {
        assert.equal(new Prng(c.seed).integer(c.key, c.range), c.result);
      });
    }
  });

  describe('shuffle', () => {
    for (const [i, c] of fixture.shuffle.entries()) {
      it(`#${i} length=${c.items.length}`, () => {
        assert.deepEqual(new Prng(c.seed).shuffle(c.key, c.items), c.result);
      });
    }
  });
});

describe('Parity / Numbers', () => {
  // The renderer formats every numeric attribute via Number.format (rounded to
  // 5 decimals); this pins that contract so the PHP/Python ports stay identical.
  for (const entry of loadFixture('numbers.json')) {
    it(`formats ${entry.output}`, () => {
      assert.equal(Number.format(entry.input), entry.output);
    });
  }
});

describe('Parity / Initials', () => {
  // Initials.fromSeed must be byte-identical across ports; accented seeds
  // (ô/ü) and quote/@/CJK/emoji cases pin the @-strip, quote-strip, and
  // \p{L} word matching against per-language regex pitfalls.
  for (const entry of loadFixture('initials.json')) {
    it(`derives ${JSON.stringify(entry.seed)} → ${JSON.stringify(entry.result)}`, () => {
      assert.equal(Initials.fromSeed(entry.seed), entry.result);
    });
  }
});

describe('Parity / Color', () => {
  const colors = loadFixture('colors.json');

  for (const entry of colors.toHex) {
    it(`toHex ${entry.input}`, () => {
      assert.equal(Color.toHex(entry.input), entry.result);
    });
  }

  for (const entry of colors.toRgbHex) {
    it(`toRgbHex ${entry.input}`, () => {
      assert.equal(Color.toRgbHex(entry.input), entry.result);
    });
  }

  for (const entry of colors.parseHex) {
    it(`parseHex ${entry.input}`, () => {
      assert.deepEqual(Color.parseHex(entry.input), entry.result);
    });
  }

  for (const entry of colors.luminance) {
    it(`luminance ${entry.input}`, () => {
      assert.equal(Color.luminance(entry.input), entry.result);
    });
  }

  for (const [i, entry] of colors.sortByContrast.entries()) {
    it(`sortByContrast #${i}`, () => {
      assert.deepEqual(
        Color.sortByContrast(entry.candidates, entry.refColor),
        entry.result,
      );
    });
  }

  for (const [i, entry] of colors.filterNotEqualTo.entries()) {
    it(`filterNotEqualTo #${i}`, () => {
      assert.deepEqual(
        Color.filterNotEqualTo(entry.candidates, entry.excluded),
        entry.result,
      );
    });
  }
});

describe('Parity / Validation', () => {
  const validation = loadFixture('validation.json');
  const minimalStyle = new Style(
    validation.styles.find((e) => e.id === 'minimal').definition,
  );

  for (const entry of validation.styles) {
    it(`style ${entry.id}`, () => {
      if (entry.valid) {
        new Style(entry.definition);
      } else {
        assert.throws(() => new Style(entry.definition), ValidationError);
      }
    });
  }

  for (const entry of validation.options) {
    it(`options ${entry.id}`, () => {
      if (entry.valid) {
        new Avatar(minimalStyle, entry.options);
      } else {
        assert.throws(
          () => new Avatar(minimalStyle, entry.options),
          ValidationError,
        );
      }
    });
  }

  for (const entry of validation.circularColors) {
    it(`circular colors ${entry.id}`, () => {
      assert.throws(
        () => new Avatar(new Style(entry.style), entry.options),
        (e) => {
          assert.ok(e instanceof CircularColorReferenceError);
          assert.deepEqual(e.chain, entry.chain);

          return true;
        },
      );
    });
  }
});

describe('Parity / OptionsDescriptor', () => {
  const styleNames = readdirSync(join(fixturesDir, 'descriptors'))
    .filter((f) => f.endsWith('.json'))
    .map((f) => f.replace(/\.json$/, ''));

  for (const styleName of styleNames) {
    it(styleName, () => {
      const styleData = loadFixture(`styles/${styleName}.json`);
      const expected = loadFixture(`descriptors/${styleName}.json`);

      assert.deepEqual(
        new OptionsDescriptor(new Style(styleData)).toJSON(),
        expected,
      );
    });
  }
});

describe('Parity / Avatar', () => {
  const styleNames = readdirSync(join(fixturesDir, 'avatars'))
    .filter((f) => f.endsWith('.json'))
    .map((f) => f.replace(/\.json$/, ''));

  for (const styleName of styleNames) {
    describe(styleName, () => {
      const styleData = loadFixture(`styles/${styleName}.json`);
      const cases = loadFixture(`avatars/${styleName}.json`);

      for (const entry of cases) {
        it(entry.id, () => {
          const avatar = new Avatar(new Style(styleData), entry.options);
          const json = avatar.toJSON();
          assert.equal(json.svg, entry.svg);
          // Round-trip via JSON to drop `undefined` values (size/title),
          // mirroring what real consumers see after JSON.stringify.
          assert.deepEqual(
            JSON.parse(JSON.stringify(json.options)),
            entry.resolvedOptions,
          );

          // Only select cases carry a dataUri — it pins the percent-encoding
          // contract (encodeURIComponent) without bloating every fixture.
          if (entry.dataUri !== undefined) {
            assert.equal(avatar.toDataUri(), entry.dataUri);
          }
        });
      }
    });
  }
});

File: src/go/core/parity_test.go
package dicebear

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	"github.com/dicebear/dicebear-go/v10/color"
	"github.com/dicebear/dicebear-go/v10/internal/initials"
	"github.com/dicebear/dicebear-go/v10/internal/num"
	"github.com/dicebear/dicebear-go/v10/internal/prng"
)

// Cross-language parity: assert the primitives produce exactly the values in the
// shared fixtures under <repo>/tests/fixtures/parity/, the same ones every other
// port runs against. The tests skip gracefully when the fixtures are absent
// (e.g. in the split dicebear-go repo).

func fixturePath(t *testing.T, name string) string {
	t.Helper()
	p := filepath.Join("..", "..", "..", "tests", "fixtures", "parity", name)
	if _, err := os.Stat(p); err != nil {
		t.Skipf("parity fixtures not available (%s)", p)
	}
	return p
}

// fixtureStyleNames enumerates the parity style fixtures instead of repeating a
// hand-maintained list, so a fixture added to tests/fixtures/parity/styles is
// picked up here the way the PHP and Python suites already pick it up. Names are
// returned sorted, so the subtest order stays stable.
func fixtureStyleNames(t *testing.T) []string {
	t.Helper()

	paths, err := filepath.Glob(filepath.Join(fixturePath(t, "styles"), "*.json"))
	if err != nil {
		t.Fatalf("list style fixtures: %v", err)
	}
	if len(paths) == 0 {
		t.Fatal("no style fixtures found")
	}

	names := make([]string, 0, len(paths))
	for _, p := range paths {
		names = append(names, strings.TrimSuffix(filepath.Base(p), ".json"))
	}
	sort.Strings(names)

	return names
}

func readFixture(t *testing.T, name string, v any) {
	t.Helper()
	b, err := os.ReadFile(fixturePath(t, name))
	if err != nil {
		t.Fatalf("read %s: %v", name, err)
	}
	if err := json.Unmarshal(b, v); err != nil {
		t.Fatalf("parse %s: %v", name, err)
	}
}

func strID(s string) string { return s }

func TestFnv1aParity(t *testing.T) {
	var cases []struct {
		Input string `json:"input"`
		Hash  uint32 `json:"hash"`
		Hex   string `json:"hex"`
	}
	readFixture(t, "fnv1a.json", &cases)

	for _, c := range cases {
		if got := prng.Hash(c.Input); got != c.Hash {
			t.Errorf("Hash(%q) = %d, want %d", c.Input, got, c.Hash)
		}
		if got := prng.Hex(c.Input); got != c.Hex {
			t.Errorf("Hex(%q) = %s, want %s", c.Input, got, c.Hex)
		}
	}
}

func TestMulberry32Parity(t *testing.T) {
	var cases []struct {
		Seed     uint32 `json:"seed"`
		Sequence []struct {
			Float float64 `json:"float"`
			State int32   `json:"state"`
		} `json:"sequence"`
	}
	readFixture(t, "mulberry32.json", &cases)

	for _, c := range cases {
		m := prng.NewMulberry32(c.Seed)
		for i, step := range c.Sequence {
			f := m.NextFloat()
			s := m.SignedState()
			if f != step.Float {
				t.Errorf("seed %d step %d: float = %v, want %v", c.Seed, i, f, step.Float)
			}
			if s != step.State {
				t.Errorf("seed %d step %d: state = %d, want %d", c.Seed, i, s, step.State)
			}
		}
	}
}

type rangeFixture struct {
	Min  float64  `json:"min"`
	Max  float64  `json:"max"`
	Step *float64 `json:"step"`
}

func (r rangeFixture) toRange() *prng.Range {
	return &prng.Range{Min: r.Min, Max: r.Max, Step: r.Step}
}

func TestPrngParity(t *testing.T) {
	var f struct {
		GetValue []struct {
			Seed, Key string
			Result    float64
		} `json:"getValue"`
		Pick []struct {
			Seed, Key string
			Items     []string
			Result    *string
		} `json:"pick"`
		WeightedPick []struct {
			Seed, Key string
			Weights   map[string]float64
			Result    *string
		} `json:"weightedPick"`
		Bool []struct {
			Seed, Key  string
			Likelihood float64
			Result     bool
		} `json:"bool"`
		Float []struct {
			Seed, Key string
			Range     rangeFixture
			Result    float64
		} `json:"float"`
		Integer []struct {
			Seed, Key string
			Range     rangeFixture
			Result    int
		} `json:"integer"`
		Shuffle []struct {
			Seed, Key string
			Items     []string
			Result    []string
		} `json:"shuffle"`
	}
	readFixture(t, "prng.json", &f)

	for _, c := range f.GetValue {
		if got := prng.New(c.Seed).GetValue(c.Key); got != c.Result {
			t.Errorf("GetValue(%q, %q) = %v, want %v", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Pick {
		got, ok := prng.Pick(prng.New(c.Seed), c.Key, c.Items, strID)
		if !matchOptional(got, ok, c.Result) {
			t.Errorf("Pick(%q, %q, %v) = (%q, %v), want %v", c.Seed, c.Key, c.Items, got, ok, c.Result)
		}
	}

	for _, c := range f.WeightedPick {
		got, ok := prng.New(c.Seed).WeightedPick(c.Key, c.Weights)
		if !matchOptional(got, ok, c.Result) {
			t.Errorf("WeightedPick(%q, %q, %v) = (%q, %v), want %v", c.Seed, c.Key, c.Weights, got, ok, c.Result)
		}
	}

	for _, c := range f.Bool {
		if got := prng.New(c.Seed).Bool(c.Key, c.Likelihood); got != c.Result {
			t.Errorf("Bool(%q, %q, %v) = %v, want %v", c.Seed, c.Key, c.Likelihood, got, c.Result)
		}
	}

	for _, c := range f.Float {
		if got := prng.New(c.Seed).Float(c.Key, c.Range.toRange()); got != c.Result {
			t.Errorf("Float(%q, %q) = %v, want %v", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Integer {
		if got := prng.New(c.Seed).Integer(c.Key, c.Range.toRange()); got != c.Result {
			t.Errorf("Integer(%q, %q) = %d, want %d", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Shuffle {
		got := prng.New(c.Seed).Shuffle(c.Key, c.Items)
		if !equalStrings(got, c.Result) {
			t.Errorf("Shuffle(%q, %q, %v) = %v, want %v", c.Seed, c.Key, c.Items, got, c.Result)
		}
	}
}

func TestNumberParity(t *testing.T) {
	var cases []struct {
		Input  float64 `json:"input"`
		Output string  `json:"output"`
	}
	readFixture(t, "numbers.json", &cases)

	for _, c := range cases {
		if got := num.Format(c.Input); got != c.Output {
			t.Errorf("num.Format(%v) = %q, want %q", c.Input, got, c.Output)
		}
	}
}

func TestInitialsParity(t *testing.T) {
	var cases []struct {
		Seed   string `json:"seed"`
		Result string `json:"result"`
	}
	readFixture(t, "initials.json", &cases)

	for _, c := range cases {
		if got := initials.FromSeed(c.Seed); got != c.Result {
			t.Errorf("initials.FromSeed(%q) = %q, want %q", c.Seed, got, c.Result)
		}
	}
}

func TestColorParity(t *testing.T) {
	var f struct {
		ToHex []struct {
			Input  string `json:"input"`
			Result string `json:"result"`
		} `json:"toHex"`
		ToRGBHex []struct {
			Input  string `json:"input"`
			Result string `json:"result"`
		} `json:"toRgbHex"`
		ParseHex []struct {
			Input  string  `json:"input"`
			Result []uint8 `json:"result"`
		} `json:"parseHex"`
		Luminance []struct {
			Input  string  `json:"input"`
			Result float64 `json:"result"`
		} `json:"luminance"`
		SortByContrast []struct {
			Candidates []string `json:"candidates"`
			RefColor   string   `json:"refColor"`
			Result     []string `json:"result"`
		} `json:"sortByContrast"`
		FilterNotEqualTo []struct {
			Candidates []string `json:"candidates"`
			Excluded   []string `json:"excluded"`
			Result     []string `json:"result"`
		} `json:"filterNotEqualTo"`
	}
	readFixture(t, "colors.json", &f)

	for _, c := range f.ToHex {
		if got := color.ToHex(c.Input); got != c.Result {
			t.Errorf("ToHex(%q) = %q, want %q", c.Input, got, c.Result)
		}
	}

	for _, c := range f.ToRGBHex {
		if got := color.ToRGBHex(c.Input); got != c.Result {
			t.Errorf("ToRGBHex(%q) = %q, want %q", c.Input, got, c.Result)
		}
	}

	for _, c := range f.ParseHex {
		r, g, b := color.ParseHex(c.Input)
		if r != c.Result[0] || g != c.Result[1] || b != c.Result[2] {
			t.Errorf("ParseHex(%q) = (%d, %d, %d), want %v", c.Input, r, g, b, c.Result)
		}
	}

	for _, c := range f.Luminance {
		if got := color.Luminance(c.Input); got != c.Result {
			t.Errorf("Luminance(%q) = %v, want %v", c.Input, got, c.Result)
		}
	}

	for _, c := range f.SortByContrast {
		if got := color.SortByContrast(c.Candidates, c.RefColor); !equalStrings(got, c.Result) {
			t.Errorf("SortByContrast(%v, %q) = %v, want %v", c.Candidates, c.RefColor, got, c.Result)
		}
	}

	for _, c := range f.FilterNotEqualTo {
		if got := color.FilterNotEqualTo(c.Candidates, c.Excluded); !equalStrings(got, c.Result) {
			t.Errorf("FilterNotEqualTo(%v, %v) = %v, want %v", c.Candidates, c.Excluded, got, c.Result)
		}
	}
}

// Cross-language validation parity: every port must accept and reject the same
// inputs (error messages are language-specific and not compared). The circular
// color reference cases additionally pin the reported chain.
func TestValidationParity(t *testing.T) {
	var f struct {
		Styles []struct {
			ID         string          `json:"id"`
			Definition json.RawMessage `json:"definition"`
			Valid      bool            `json:"valid"`
		} `json:"styles"`
		Options []struct {
			ID      string         `json:"id"`
			Options map[string]any `json:"options"`
			Valid   bool           `json:"valid"`
		} `json:"options"`
		CircularColors []struct {
			ID      string          `json:"id"`
			Style   json.RawMessage `json:"style"`
			Options map[string]any  `json:"options"`
			Chain   []string        `json:"chain"`
		} `json:"circularColors"`
	}
	readFixture(t, "validation.json", &f)

	var minimal *Style
	for _, c := range f.Styles {
		s, err := NewStyle(c.Definition)
		if c.Valid != (err == nil) {
			t.Errorf("style %s: err = %v, want valid = %v", c.ID, err, c.Valid)
			continue
		}
		if !c.Valid {
			var ve *ValidationError
			if !errors.As(err, &ve) {
				t.Errorf("style %s: error %T is not a *ValidationError", c.ID, err)
			}
		}
		if c.ID == "minimal" {
			minimal = s
		}
	}
	if minimal == nil {
		t.Fatal("minimal style fixture missing or invalid")
	}

	for _, c := range f.Options {
		_, err := NewAvatar(minimal, c.Options)
		if c.Valid != (err == nil) {
			t.Errorf("options %s: err = %v, want valid = %v", c.ID, err, c.Valid)
			continue
		}
		if !c.Valid {
			var ve *ValidationError
			if !errors.As(err, &ve) {
				t.Errorf("options %s: error %T is not a *ValidationError", c.ID, err)
			}
		}
	}

	for _, c := range f.CircularColors {
		style, err := NewStyle(c.Style)
		if err != nil {
			t.Errorf("circular %s: style must parse: %v", c.ID, err)
			continue
		}
		_, err = NewAvatar(style, c.Options)
		var ce *CircularColorReferenceError
		if !errors.As(err, &ce) {
			t.Errorf("circular %s: err = %v, want *CircularColorReferenceError", c.ID, err)
			continue
		}
		if !equalStrings(ce.Chain, c.Chain) {
			t.Errorf("circular %s: chain = %v, want %v", c.ID, ce.Chain, c.Chain)
		}
	}
}

func matchOptional(got string, ok bool, want *string) bool {
	if want == nil {
		return !ok
	}
	return ok && got == *want
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
